dramatiq/middleware/retries.py

```diff
@@ -114,3 +115,9 @@ def after_process_message(self, broker, message, *, result=None, exception=None)
 
         self.logger.info("Retrying message %r in %d milliseconds.", message.message_id, delay)
         broker.enqueue(message, delay=delay)
+
+    def before_enqueue(self, broker, message, delay):
+
+        # If message will be retried, record in its options time at which it is requeued
+        if "retries" in message.options:
+            message.options["requeue_timestamp"] = int(time.time() * 1000)
```

Comment:
Since you’re changing the Retries middleware directly, I would move this line under line 98 and avoid the check and separate before enqueue hook altogether.

Reply:
My small concern was that maybe, just maybe, in corner cases there might be some lag between execution of that part of code and enqueue operation, given how dramatiq might be running workers doing heavy CPU operations at the same time.
I shouldn't worry about that?

Reply:
I think the most likely place for this thread to relinquish control is going to be the io that is going to be done when actually enqeueuing the message so I think the chances of the suggested approach being less reliable for timing are very slim (it might in fact be more reliable because it’s doing less work).

Reply:
Roger that, I'll refactor and push again 🫡